### PR TITLE
Add broadcast semantics

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 1]
```diff
@@ -416,7 +416,13 @@ bool isTppAdd(linalg::GenericOp linalgOp) {
     return false;
   if (!isBinaryOp(linalgOp))
     return false;
-  if (!allOperandsHaveSameShapeAndStrides(linalgOp->getOperands().getTypes()))
+  if (!allOperandsHaveSameShapeAndStrides(linalgOp->getOperands().getTypes()) &&
+      (verifyTppIdentityBroadcastingRules(linalgOp.getOperand(1).getType(),
```
Comment (chelini) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:420]: why `verifyTppIdentityBroadcastingRules`? You can just use `verifyBroadcastableShapeImpl` (see: https://github.com/plaidml/tpp-mlir/commit/e8002870ba5115d6885911b0eff0e24d8ce1f79d).

**Comment by rengolin** [comment]
Hopefully, with the new accumulation type fix, this will be green. If not, the dump @adam-smnk introduced should tell us what's wrong.

**Review by rengolin — APPROVED** [review]

File: test/Conversion/LinalgToTpp/linalg-to-tpp.mlir [hunk 2]
```diff
@@ -64,12 +64,11 @@ func.func @add_mapping_scalar(%arg0: memref<f32>, %arg1: memref<f32>) {
 }
 
 // -----
-
 // We don't support broadcast for tpp.add. All operands must have the same type.
 #map = affine_map<(d0, d1) -> (d0, d1)>
 #map1 = affine_map<(d0, d1) -> (0, d1)>
 func.func @add_mapping_brcst(%arg0: memref<3x3xf32>, %arg1: memref<1x3xf32>) {
-  // CHECK-NOT: tpp.add
+// expected-error @below {{'linalg.generic' op result type not broadcast compatible with broadcasted operands's shapes}}
```
Comment (chelini) [inline test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: we should not emit diagnostic here.

Comment (rengolin) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: Good point. It should just be lowered to non-TPP code.

Comment (KavithaTipturMadhu) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: I thought the diagnostic came from verifyBroadcastableShapeImpl? I didn't introduce it obviously, I just use the check verifyBroadcastableShapeImpl in isTppAdd util. Is there a way to supress the diagnostic?

Comment (chelini) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: yes, see: checkBroadcastableShape in https://github.com/plaidml/tpp-mlir/pull/424. If it is urgent I can try to split the PR.

Comment (KavithaTipturMadhu) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: Its not urgent, we can wait till the other PR lands, no problem.

Comment (rengolin) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: Can we bypass the other PR to merge this first? IIUC, this is causing a performance regression on the benchmarks as it's not detecting the bias add and lowers to loops.

Comment (chelini) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:71]: We converged on the matcher design. I am planning to work on the PR now and hopefully merge it soon. If we want to land this first, we can take the changes for `checkBroadcastableShape`? Preference?

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 3]
```diff
@@ -338,6 +330,15 @@ static bool hasReluBody(Operation *op, SmallVectorImpl<Value> *captured) {
   return (getOperand(maxfLhs, maxfRhs) || getOperand(maxfRhs, maxfLhs));
 }
 
+// Return true if the linalg.generic can be mapped to a tpp.add.
+bool isTppAdd(linalg::GenericOp linalgOp, SmallVectorImpl<Value> *operands) {
+  using namespace tpp::structured_match;
+  auto addMatcher = StructuredOpMatcher::make<linalg::GenericOp>().region(
+      WithSingleOp<arith::AddFOp>(), operands);
+  auto res = mlir::OpTrait::tpp::checkBroadcastableShape(linalgOp);
```
Comment (chelini) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:338]: no need for this check. Is already done in `isTppBinaryOp`

**Pull request closed without merging**